"AssertionError: Out of order violation for key ..." for counting window

#### Is this a bug, feature request, or feedback?

Bug

#### What is the current behavior?

Intermittent failure, such as:
```
cd /home/vagrant/wallaroo/testing/correctness/apps/window_detector/ && \
integration_test \
	--alo-sequence-sender key_0 '(0,100]' Detector \
	--alo-sequence-sender key_1 '(0,100]' Detector \
	--alo-sequence-sender key_0 '(100,200]' Detector \
	--alo-sequence-sender key_1 '(100,200]' Detector \
	--log-level error \
	--command 'machida --application-module window_detector --source alo --window-type counting ' \
	--validation-cmd 'python _validate.py --window-type counting --output' \
	--output 'received.txt' \
	--batch-size 50 \
	--workers 5 \
	--sink-await-key key_0 200 \
	--sink-await-key key_1 200
2019-07-24 02:15:39,088 integration ERROR    [cluster.py:1042] An error was raised in the Cluster context
Traceback (most recent call last):
  File "/home/vagrant/wallaroo/testing/tools/integration/integration.py", line 300, in pipeline_test
    raise PipelineTestError(error)
integration.errors.PipelineTestError: Validation command '['python', '_validate.py', '--window-type', 'counting', '--output', '/tmp/res-data.zn1x_gt3/sink.0.dat']' failed with the output:
--
Traceback (most recent call last):
  File "_validate.py", line 49, in <module>
    .format(k, w_key, win, sorted(win)))
AssertionError: Out of order violation for key: key_0, w_key: 1563934538.84, window: [121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70], sorted: [31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130]
```
#### What is the expected behavior?

Test always passes

#### What OS and version of Wallaroo are you using?

Ubuntu Xenial 64-bit, 1 or 2 virtual CPU, 4GB RAM, `simplify-connectors-1` branch at commit e316b01b4, which is based at `master` branch f748f92efe.

#### Steps to reproduce?

Apply patch at https://gist.github.com/slfritchie/6c2a9c81467fe884b5c984f6698e663e, then:

```
cd testing/correctness/apps/window_detector
make test debug=true PONYCFLAGS="--verbose=1 -Dtrace -Dcheckpoint_trace -Didentify_routing_ids" && ../../../tools/try_until/try_until 999 make test debug=true
export WALLAROO_THRESHOLDS="*.8"
make test debug=true 'PONYCFLAGS="--verbose=1 -Dtrace -Dcheckpoint_trace -Didentify_routing_ids"' && ../../../tools/try_until/try_until 999 make test debug=true
```

Full logs are at http://wallaroolabs-dev.s3.amazonaws.com/logs/test-artifacts.b.20190723.tar.gz

* All messages for key 0 arrive in the correct order at the initializer worker.
* Key 0's messages appear at worker4, out of order, when printing messages such as `trace_id((key_0,1))`.
     * I don't know if it's relevant, but the first of these `trace_id()` messages are printed in the middle of a "unmute storm" similar to bug #2912 
* Key 0's messages appear at worker3, out of order, when printing messages such as `!@ Collect.update` and `Collect.output`.

I'm 95% sure that the root of this problem lies in the `initializer` worker.

If I use the `parallelism=1` setting for the ALO source (see commit 6a2e6d78), then this problem disappears.  (This bug is intermittent, but it typically appears at least once per hour when I test on a 2 virtual CPU VM.  With `parallelism=1`, I have seen 0 bug instances in over 24 hours.)

When source `parallelism` is greater than 1, then we have a problem that Wallaroo today cannot solve.  All instances of the `AssertionError: Out of order violation for key` error involve messages with values on both "sides" of the value 100.  Note that the test uses 2 different ALO source connections to send values 1-100 and 101-200 to the ALO source.  A deterministic hash is used to map the key to the connector protocol `stream_id`; the same `stream_id` is used by both TCP connections to the Wallaroo connector source.

However, during an intermittent failing run, the two TCP connections to the Wallaroo connector source are handled by different `ConnectorSource` actors.  Due to the nature of the app pipeline definition, each `ConnectorSource` actor @ the head of the pipeline will use different `OutgoingBoundary` actors to forward messages from `initializer` to `worker1`.  Wallaroo today has no synchronization mechanism to preserve order for messages in the same `stream_id` that also flow through different `OutgoingBoundary` actors.  All values in the 1-100 range and in the 101-200 range are always being delivered in the correct order.  However, the intermittent failures always manage to interleave messages from both ranges, which is a no-no.

Fixes to this bug can include:

* Change the nature of the test to avoid multiple connections for the same stream ID
* Remove the test altogether
* Add a barrier to be used whenever a connector protocol `stream_id` is re-used (i.e. 2nd and following `Notify` messages for a `stream_id`)
    * Alternative: use the existing checkpoint barrier (plus its accompanying side effects) as the sync method

Thoughts @jtfmumm, @SeanTAllen?

--- a/testing/correctness/apps/window_detector/Makefile
+++ b/testing/correctness/apps/window_detector/Makefile
@@ -134,48 +134,42 @@ window_detector_tumbling_alo_source_test_python:
 	integration_test \
 		--alo-sequence-sender key_0 '(0,100]' Detector \
 		--alo-sequence-sender key_1 '(0,100]' Detector \
-		--alo-sequence-sender key_0 '(100,200]' Detector \
-		--alo-sequence-sender key_1 '(100,200]' Detector \
 		--log-level $(LOGLEVEL) \
 		--command 'machida --application-module window_detector --source alo --window-type tumbling --window-late-policy fire-per-message $(RUN_WITH_RESILIENCE)' \
 		--validation-cmd 'python _validate.py --window-type tumbling --output' \
 		--output 'received.txt' \
 		--batch-size 50 \
 		--workers 5 \
-		--sink-await-key key_0 200 \
-		--sink-await-key key_1 200
+		--sink-await-key key_0 100 \
+		--sink-await-key key_1 100
 
 window_detector_sliding_alo_source_test_python:
 	cd $(WINDOW_DETECTOR_PATH) && \
 	integration_test \
 		--alo-sequence-sender key_0 '(0,100]' Detector \
 		--alo-sequence-sender key_1 '(0,100]' Detector \
-		--alo-sequence-sender key_0 '(100,200]' Detector \
-		--alo-sequence-sender key_1 '(100,200]' Detector \
 		--log-level $(LOGLEVEL) \
 		--command 'machida --application-module window_detector --source alo --window-type sliding --window-late-policy fire-per-message $(RUN_WITH_RESILIENCE)' \
 		--validation-cmd 'python _validate.py --window-type sliding --output' \
 		--output 'received.txt' \
 		--batch-size 50 \
 		--workers 5 \
-		--sink-await-key key_0 200 \
-		--sink-await-key key_1 200
+		--sink-await-key key_0 100 \
+		--sink-await-key key_1 100
 
 window_detector_counting_alo_source_test_python:
 	cd $(WINDOW_DETECTOR_PATH) && \
 	integration_test \
 		--alo-sequence-sender key_0 '(0,100]' Detector \
 		--alo-sequence-sender key_1 '(0,100]' Detector \
-		--alo-sequence-sender key_0 '(100,200]' Detector \
-		--alo-sequence-sender key_1 '(100,200]' Detector \
 		--log-level $(LOGLEVEL) \
 		--command 'machida --application-module window_detector --source alo --window-type counting $(RUN_WITH_RESILIENCE)' \
 		--validation-cmd 'python _validate.py --window-type counting --output' \
 		--output 'received.txt' \
 		--batch-size 50 \
 		--workers 5 \
-		--sink-await-key key_0 200 \
-		--sink-await-key key_1 200
+		--sink-await-key key_0 100 \
+		--sink-await-key key_1 100
 
 window_detector_tumbling_test_python:
 	cd $(WINDOW_DETECTOR_PATH) && \
@@ -261,48 +255,42 @@ window_detector_tumbling_alo_source_test_python3:
 	integration_test \
 		--alo-sequence-sender key_0 '(0,100]' Detector \
 		--alo-sequence-sender key_1 '(0,100]' Detector \
-		--alo-sequence-sender key_0 '(100, 200]' Detector \
-		--alo-sequence-sender key_1 '(100, 200]' Detector \
 		--log-level $(LOGLEVEL) \
 		--command 'machida3 --application-module window_detector --source alo --window-type tumbling --window-late-policy fire-per-message $(RUN_WITH_RESILIENCE)' \
 		--validation-cmd 'python _validate.py --window-type tumbling --output' \
 		--output 'received.txt' \
 		--batch-size 50 \
 		--workers 5 \
-		--sink-await-key key_0 200 \
-		--sink-await-key key_1 200
+		--sink-await-key key_0 100 \
+		--sink-await-key key_1 100
 
 window_detector_sliding_alo_source_test_python3:
 	cd $(WINDOW_DETECTOR_PATH) && \
 	integration_test \
 		--alo-sequence-sender key_0 '(0,100]' Detector \
 		--alo-sequence-sender key_1 '(0,100]' Detector \
-		--alo-sequence-sender key_0 '(100, 200]' Detector \
-		--alo-sequence-sender key_1 '(100, 200]' Detector \
 		--log-level $(LOGLEVEL) \
 		--command 'machida3 --application-module window_detector --source alo --window-type sliding --window-late-policy fire-per-message $(RUN_WITH_RESILIENCE)' \
 		--validation-cmd 'python _validate.py --window-type sliding --output' \
 		--output 'received.txt' \
 		--batch-size 50 \
 		--workers 5 \
-		--sink-await-key key_0 200 \
-		--sink-await-key key_1 200
+		--sink-await-key key_0 100 \
+		--sink-await-key key_1 100
 
 window_detector_counting_alo_source_test_python3:
 	cd $(WINDOW_DETECTOR_PATH) && \
 	integration_test \
 		--alo-sequence-sender key_0 '(0,100]' Detector \
 		--alo-sequence-sender key_1 '(0,100]' Detector \
-		--alo-sequence-sender key_0 '(100, 200]' Detector \
-		--alo-sequence-sender key_1 '(100, 200]' Detector \
 		--log-level $(LOGLEVEL) \
 		--command 'machida3 --application-module window_detector --source alo --window-type counting $(RUN_WITH_RESILIENCE)' \
 		--validation-cmd 'python _validate.py --window-type counting --output' \
 		--output 'received.txt' \
 		--batch-size 50 \
 		--workers 5 \
-		--sink-await-key key_0 200 \
-		--sink-await-key key_1 200
+		--sink-await-key key_0 100 \
+		--sink-await-key key_1 100
 
 window_detector_tumbling_test_python3:
 	cd $(WINDOW_DETECTOR_PATH) && \

